# DS-SF-34 | 02 | The `pandas` Library | Assignment | Starter Code

## The Bistro Meets `pandas`

You've just told one of your friend that you are taking a Data Science class.  (Yeah!)  Your friend is running a bistro, a small restaurant, serving moderately priced simple meals in a modest setting ([Wikipedia](https://en.wikipedia.org/wiki/Bistro)).  She collected over some period of time the following information of her patrons' visits.

| Variable's name | Its meaning |
|:---:|:---|
| `name` | Patron's first name |
| `gender` | Patron's gender |
| `is_smoker` | Whether the patron is smoking or not |
| `party` | Party's size |
| `check` | Check amount (\$) (after taxes but before tips) |
| `tip` | Tip (\$) that the patron added to the check |
| `day` | Week day of the visit |
| `time` | Rough time estimate of the visit |

In this assignment, we will be exploring this dataset using `pandas`.<sup>(*)</sup>

<sup>(*)</sup> this dataset was adapted from the `tips` dataset of the `seaborn` package (https://github.com/mwaskom/seaborn-data)

> ### Question 1.  Import `numpy` (as `np`) and `pandas` (as `pd`).

In [1]:
import os
import pandas as pd
import numpy as np

# TODO

pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

> ### Question 2.  Read the `dataset-02-tips.csv` dataset.

In [2]:
# TODO
df = pd.read_csv('../datasets/dataset-02-bistro.csv')
df.head()

day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66
2  Sunday  Dinner     Larry    Male      False      3  21.01  3.50
3  Sunday  Dinner    Joseph    Male      False      2  23.68  3.31
4  Sunday  Dinner    Janice  Female      False      4  24.59  3.61

> ### Question 3.  What is the class of the `pandas` object storing the dataset?

In [3]:
# TODO
type(df)

pandas.core.frame.DataFrame

Answer: TODO

> ### Question 4.  How many samples (i.e., rows) are in this dataset?

In [4]:
# TODO rows in this dataset
#df.shape[0]
# or 
len(df)

244

Answer: TODO

> ### Question 5.  How many variables (i.e., columns) are in this dataset?

In [5]:
# TODO
df.shape[1]
#or 
len(df.columns)

8

Answer: TODO

> ### Question 6.  Print the name of each column in the dataset, one name per line.

In [6]:
# TODO No need of a list
for column in df.columns:
    print(column)


day
time
name
gender
is_smoker
party
check
tip


> ### Question 7.  Print the first two rows of the dataset to the console.  What does the output look like?

In [9]:
# TODO
print(df.head(2))
print(df[:2])

      day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66
      day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66


Answer: TODO

> ### Question 8.  Extract the last 2 rows of the data frame and print them to the console.  What does the output look like?

In [10]:
# TODO
df.tail(2)

day    time    name  gender  is_smoker  party  check   tip
242  Saturday  Dinner     Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner  Brandi  Female      False      2  18.78  3.00

In [11]:
#or
df[-2:]

day    time    name  gender  is_smoker  party  check   tip
242  Saturday  Dinner     Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner  Brandi  Female      False      2  18.78  3.00

In [12]:
df.iloc[-2:]

day    time    name  gender  is_smoker  party  check   tip
242  Saturday  Dinner     Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner  Brandi  Female      False      2  18.78  3.00

Answer: TODO

> ### Question 9.  Does the dataset contain any missing values?

In [13]:
df.isnull().sum().sum()

0

In [14]:


# TODO
df.isnull().any() # no it doesn't have any missing values

day          False
time         False
name         False
gender       False
is_smoker    False
party        False
check        False
tip          False
dtype: bool

Answer: TODO

> ### Question 10.  What can you say about the `is_smoker` variable?  I.e., will it bring any insights when analyzing the dataset?  What do you want to do with it?  (and do it...)

In [15]:
# TODO
df.is_smoker.describe()
# All values are false so we can remove it.

count       244
unique        1
top       False
freq        244
Name: is_smoker, dtype: object

In [16]:
df = df.drop('is_smoker', axis = 1)

Answer: All values are false so we can remove it.

> ### Question 11.  For which week days does the dataset has data for?

In [17]:
# TODO
df.day.unique()

array(['Sunday', 'Saturday', 'Thursday', 'Friday'], dtype=object)

Answer: TODO

> ### Question 12.  How often was the bistro patronized for each week day?

(check `.value_counts()`; it could come in handy)

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html)

In [18]:
# TODO
df.day.value_counts()

Saturday    87
Sunday      76
Thursday    62
Friday      19
Name: day, dtype: int64

Answer: TODO

In [20]:
df.groupby('day')['tip'].count()

day
Friday      19
Saturday    87
Sunday      76
Thursday    62
Name: tip, dtype: int64

> ### Question 13.  How much tip did waiters collect for each week day?

In [14]:
# TODO
tip_collected = df.groupby('day')['tip'].sum()
print(tip_collected)

day
Friday       51.96
Saturday    260.40
Sunday      247.39
Thursday    171.83
Name: tip, dtype: float64


Answer: TODO

> ### Question 14.  What is the average tip per check (in absolute \$) for each week day?

In [21]:
# TODO
df.groupby('day')['tip'].mean()


day
Friday      2.734737
Saturday    2.993103
Sunday      3.255132
Thursday    2.771452
Name: tip, dtype: float64

In [22]:
df.groupby('day')['tip'].sum()/df.groupby('day')['tip'].count()

day
Friday      2.734737
Saturday    2.993103
Sunday      3.255132
Thursday    2.771452
Name: tip, dtype: float64

Answer: TODO

> ### Question 15.  What is the average tip per check (as a percentage of the check) for each week day?

In [24]:
# TODO
Percentage_tip = (df.groupby('day')['tip'].sum()/df.groupby('day')['check'].sum())*100
print(Percentage_tip,'%')

day
Friday      15.944519
Saturday    14.642375
Sunday      15.203791
Thursday    15.673201
dtype: float64 %


Answer: TODO

> ### Question 16.  Are there any name in common between male and female patrons?  (E.g., `Chris` can refer to either a man or a woman)

(check `numpy.intersect1d()`; it could come in handy)

(https://docs.scipy.org/doc/numpy/reference/generated/numpy.intersect1d.html)

In [25]:
np.intersect1d(df[df['gender'] == 'Male'].name,df[df['gender'] == 'Female'].name)

array(['Casey'], dtype=object)

In [20]:
# TODO
df.head()


day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66
2  Sunday  Dinner     Larry    Male      False      3  21.01  3.50
3  Sunday  Dinner    Joseph    Male      False      2  23.68  3.31
4  Sunday  Dinner    Janice  Female      False      4  24.59  3.61

Answer: TODO

> ### Question 17.  If no patrons share the same name, how many unique patrons are in the dataset?

In [26]:
# TODO
len(df[df['gender']=='Male'].name.unique()) + len(df[df['gender'] == 'Female'].name.unique())

182

In [29]:
len(df.groupby(['name','gender']))

182

Answer: TODO

> ### Question 18.  How many times did `Kevin` patronized the bistro?  How about `Alice`?

In [24]:
# TODO
df_1 = df[df.name == 'Kevin']
df_1.name.value_counts()

Kevin    4
Name: name, dtype: int64

In [31]:
(df.name == 'Alice').sum()

2

Answer: TODO

> ### Question 19.  Who are the top 3 female and male patrons?

In [37]:
# TODO
s = df.groupby('gender')['name'].value_counts()

print(s.groupby(level=0).nlargest(3))

gender  gender  name 
Female  Female  Mary     4
                Casey    3
                Laura    3
Male    Male    David    8
                Casey    5
                James    5
Name: name, dtype: int64


Answer: TODO

> ### Question 20.  Who's the best tipper (as a fraction of all tips over all check totals)?  Who's the worst?  How many times did they patronize the bistro?

(check `numpy.intersect1d()`; it could come in handy)

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.sort_values.html)

In [34]:
# TODO
(df.groupby(['name','gender']).tip.sum()/df.groupby(['name','gender']).check.sum()).sort_values(ascending = False)

name      gender
Maryann   Female    0.416667
Bailey    Female    0.325733
Daniel    Male      0.313607
Dennis    Male      0.280535
Zackary   Male      0.266312
                      ...   
Willie    Male      0.066534
Kimberly  Female    0.059447
Destiny   Female    0.056797
Mildred   Female    0.056433
Jeremy    Male      0.035638
Length: 182, dtype: float64

Answer: TODO

In [35]:
df[df['name'] == 'Maryann']

day    time     name  gender  party  check  tip
178  Sunday  Dinner  Maryann  Female      2    9.6  4.0

In [36]:
df[df['name'] == 'Jeremy']

day    time    name gender  party  check   tip
237  Saturday  Dinner  Jeremy   Male      2  32.83  1.17

Maryann tipped 4.0 on a 9.6$ check and she came only once to the bistro
Jeremy tipped 1.17 on a 32.83$ check which is worst and he also came only once to the bistro